### PR TITLE
reduce loop structure leakage into the `TrainingEpochLoop`

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

This refactor aims at simplifying the output processing of our nested loop structure. The processing mainly revolves around the nested list structure that the `training_epoch_end(self, outputs)` hook would receive. Until now, the processing was scattered in different places and loops assumed knowledge of each other's output format. We refactor the following components:

- each loop obtains strict typing of its output list
- `OptimizerLoop`: Now returns a dict, indexed by the optimizer_idx that was active
- `TrainingBatchLoop`: mostly unchanged, only typing updates
- `TrainingEpochLoop`: No longer assumes knowledge of how many optimizers were running in OptimizerLoop. It only maintains a list of batch outputs and accumulates them in a list (if required).

As an intermediate step, we modify the processing in the `TrainingEpochLoop` such that the old format that the `training_epoch_end` hook accepts is still respected. No breaking changes intended. 
However, the "old" format of the output structure does not correspond to the actual loop structure. The choice of how the array list is layed out to the user is an arbitrary one. This unintuitive output format only matters when we have multiple optimizers or `truncated_bptt_steps > 1` and in the future we would like to standardize this such that the majority of our output processing can be completely removed. 

co-author: @awaelchli

### Does your PR introduce any breaking changes? If yes, please list them.

None user-facing intended

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

File: pytorch_lightning/loops/epoch/training_epoch_loop.py [hunk 1]
```diff
@@ -266,76 +263,66 @@ def _should_accumulate(self) -> bool:
...
 
-    def _track_epoch_end_reduce_metrics(
-        self, epoch_output: List[List[STEP_OUTPUT]], batch_end_outputs: STEP_OUTPUT
-    ) -> None:
-        """Adds the batch outputs to the epoch outputs and prepares reduction."""
-        hook_overridden = is_overridden("training_epoch_end", self.trainer.lightning_module)
-        if not hook_overridden:
-            return
+    @staticmethod
+    def _swap_optimizers_and_tbptt(tbptt_outputs: _BATCH_OUTPUTS_TYPE, n_opt: int) -> List[List[Dict[str, Any]]]:
```
Comment (tchaton) [inline pytorch_lightning/loops/epoch/training_epoch_loop.py:267]: Do we want to move all tbptt utilties to their own folder, same for Lighting hooks ?

Comment (carmocca) [reply pytorch_lightning/loops/epoch/training_epoch_loop.py:267]: > Do we want to move all tbptt utilties to their own folder

No because this case is a particularity of the `training_epoch_end(outputs)` format, not of `tbptt` itself. For that reason, this PR keeps this processing logic encapsulated in these `staticmethods` which ideally would be removed in the future.

File: pytorch_lightning/loops/epoch/training_epoch_loop.py [hunk 2]
```diff
@@ -146,18 +145,16 @@ def advance(self, *args: Any, **kwargs: Any) -> None:
...
             self.update_lr_schedulers("epoch", update_plateau_schedulers=False)
 
-        batch_end_outputs = [opt_idx_out for opt_idx_out in batch_output.training_step_output if len(opt_idx_out)]
-        processed_batch_end_outputs = self._prepare_outputs(batch_end_outputs, batch_mode=True)
-
-        # hook
-        self.trainer.call_hook("on_train_batch_end", processed_batch_end_outputs, batch, self.batch_idx, 0)
+        if is_overridden("on_train_batch_end", self.trainer.lightning_module):
+            batch_end_outputs = self._prepare_outputs_training_batch_end(batch_output.outputs)
+            self.trainer.call_hook("on_train_batch_end", batch_end_outputs, batch, self.batch_idx, 0)
```
Comment (awaelchli) [inline pytorch_lightning/loops/epoch/training_epoch_loop.py:150]: I think we need to call this without the overridden check, because the checkpoint callback for example relies on this. do you remember why we did this?

Comment (carmocca) [reply pytorch_lightning/loops/epoch/training_epoch_loop.py:150]: Ahh good eye.

We did this to avoid the computation when it's not overridden

But forgot that `call_hook` calls everything.

I can try and address this with #9029 - we can undo the check here.

Comment (carmocca) [reply pytorch_lightning/loops/epoch/training_epoch_loop.py:150]: It's unfortunate that `ModelCheckpoint` uses `on_train_batch_end` since it doesn't use the outputs at all. This would be one pro for #8731

File: pytorch_lightning/loops/epoch/training_epoch_loop.py [hunk 3]
```diff
@@ -268,65 +265,76 @@ def _should_accumulate(self) -> bool:
...
+        This processing is necessary because the format of the inputs to the ``training_epoch_end`` hook does not
+        match the loop structure and because empty dimensions are squeezed. This could break with loop customization.
         """
-        processed_outputs = []
-        for opt_outputs in outputs:
-            # handle an edge case where an optimizer output is the empty list
-            if len(opt_outputs) == 0:
+        # `batch_outputs` (plural) is the same as `epoch_end_output` (singular)
+        if not batch_outputs:
+            return {}
```
Comment (awaelchli) [inline pytorch_lightning/loops/epoch/training_epoch_loop.py:307]: we return a dict here but I don't remember why. the type hint says we can but I don't think that's correct.

File: pytorch_lightning/loops/epoch/training_epoch_loop.py [hunk 4]
```diff
@@ -268,65 +275,83 @@ def _should_accumulate(self) -> bool:
...
-        outputs: List[List[List[OutputResult]]], batch_mode: bool
-    ) -> Union[List[List[List[Dict]]], List[List[Dict]], List[Dict], Dict]:
-        """Extract required information from batch or epoch end results.
-
-        Args:
-            outputs: A 3-dimensional list of ``OutputResult`` objects with dimensions:
-                ``[optimizer outs][batch outs][tbptt steps]``.
-
-            batch_mode: If True, ignore the batch output dimension.
+    def _prepare_outputs_training_batch_end(
```
Comment (awaelchli) [inline pytorch_lightning/loops/epoch/training_epoch_loop.py:279]: @carmocca this code here is now very similar to the one for epoch_end. The only difference is that we have one dimension less and that the final sanitation is different. 

What do you prefer? Merge the functions into one or keep as is?

Comment (carmocca) [reply pytorch_lightning/loops/epoch/training_epoch_loop.py:279]: I'd keep them separate, in case the format for a hook changes

Comment (awaelchli) [reply pytorch_lightning/loops/epoch/training_epoch_loop.py:286]: sounds good

File: pytorch_lightning/loops/epoch/training_epoch_loop.py [hunk 5]
```diff
@@ -211,24 +216,28 @@ def on_run_end(self) -> None:
...
+        if is_overridden("training_epoch_end", model) and self._outputs:
+            epoch_end_outputs = self._prepare_outputs_training_epoch_end(
+                self._outputs,
+                automatic=model.automatic_optimization,
+                num_optimizers=len(self.trainer.optimizers),
+            )
             # check that the dataloader/iterator produced a batch
-            if processed_outputs:
+            # FIXME: still necessary?
+            if epoch_end_outputs:
```
Comment (awaelchli) [inline pytorch_lightning/loops/epoch/training_epoch_loop.py:227]: @carmocca should we make a separate PR for this? otherwise the diff here gets destroyed :)

Comment (carmocca) [reply pytorch_lightning/loops/epoch/training_epoch_loop.py:226]: For what exactly? The variable rename? The diff looks okay to me

Comment (awaelchli) [reply pytorch_lightning/loops/epoch/training_epoch_loop.py:226]: the fixme, which I interpret as "is this if statement still necessary".

Comment (carmocca) [reply pytorch_lightning/loops/epoch/training_epoch_loop.py:226]: ah sure, we can check that in a separate PR

Comment (carmocca) [reply pytorch_lightning/loops/epoch/training_epoch_loop.py:226]:
```suggestion
            if epoch_end_outputs:
```

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/loops/batch/training_batch_loop.py [hunk 6]
```diff
@@ -112,18 +115,17 @@ def advance(self, batch, batch_idx):
         # let logger connector extract current batch size
         self.trainer.logger_connector.on_train_split_start(batch_idx, split_idx, split_batch)
 
+        # choose which loop will run the optimization
         if self.trainer.lightning_module.automatic_optimization:
-            # in automatic optimization, hand over execution to the OptimizerLoop
             optimizers = _get_active_optimizers(self.trainer.optimizers, self.trainer.optimizer_frequencies, batch_idx)
```
Comment (tchaton) [inline pytorch_lightning/loops/batch/training_batch_loop.py:120]: We can make get_active_optimizers a static_method ?

Comment (tchaton) [reply pytorch_lightning/loops/batch/training_batch_loop.py:120]: Users might want to override to implement custom scheduling strategy.

**Review by tchaton — APPROVED** [review]
LGTM ! Great work !

File: pytorch_lightning/loops/batch/training_batch_loop.py (hunk 6 above)
Comment (tchaton) [reply pytorch_lightning/loops/batch/training_batch_loop.py:120]: Ignore it if you want as not fully relevant to this PR.

Comment (carmocca) [reply pytorch_lightning/loops/batch/training_batch_loop.py:120]: It could be a good idea. I guess it was put in utilities to avoid traversing the loops for it because the `EpochLoop`, `BatchLoop` and `FinetuniningCallback` all use it.

cc @awaelchli 

But definitely not in this PR

Comment (awaelchli) [reply pytorch_lightning/loops/batch/training_batch_loop.py:120]: Yes, as @carmocca said I moved `_get_active_optimizers` out of the loops because it was used in different loops and defined on the TrainingEpochLoop. But if it becomes a method, IMO it should be in the class that owns the optimizers

File: pytorch_lightning/loops/epoch/training_epoch_loop.py [hunk 7]
```diff
@@ -381,6 +394,71 @@ def _save_loggers_on_train_batch_end(self) -> None:
...
+    """Calculates the max size in each dimension of a jagged (non-rectangular) nested list of lists.
+
+    Example::
+        >>> _get_max_shape([[], [[1], [2]], []])
+        [3, 2, 1]
+    """
+    dimensions = defaultdict(int)
+    for level, length in _get_dimensions(array):
+        dimensions[level] = max(dimensions[level], length)
+    return [value for _, value in sorted(dimensions.items())]
```
Comment (rohitgr7) [inline pytorch_lightning/loops/epoch/training_epoch_loop.py:456]:
```suggestion
    return sorted(dimensions.values())
```

File: pytorch_lightning/loops/epoch/training_epoch_loop.py [hunk 8]
```diff
@@ -381,6 +394,71 @@ def _save_loggers_on_train_batch_end(self) -> None:
...
+    """Converts an optimizer dict to a list in which the key of the dict determines the position of the element.
+
+    Example::
+        >>> _convert_optim_dict({0: {"loss": 0.0}, 2: {"loss": 0.2}}, num_optimizers=3)
+        [{'loss': 0.0}, None, {'loss': 0.2}]
+    """
+    return [outs[opt_idx] if opt_idx in outs else None for opt_idx in range(num_optimizers)]
+
+
+def _recursive_unpad(nested: List[Any], value: Optional[Any] = None) -> List:
```
Comment (rohitgr7) [inline pytorch_lightning/loops/epoch/training_epoch_loop.py:408]: should nested be marked as Any since in recursive call there is a possibility of not getting a list??

File: pytorch_lightning/loops/epoch/training_epoch_loop.py [hunk 9]
```diff
@@ -381,6 +394,71 @@ def _save_loggers_on_train_batch_end(self) -> None:
...
+        >>> _get_max_shape([[], [[1], [2]], []])
+        [3, 2, 1]
+    """
+    dimensions = defaultdict(int)
+    for level, length in _get_dimensions(array):
+        dimensions[level] = max(dimensions[level], length)
+    return [value for _, value in sorted(dimensions.items())]
+
+
+def _iterate_nested_array(array: List[Any], index: Tuple = ()) -> Generator:
```
Comment (rohitgr7) [inline pytorch_lightning/loops/epoch/training_epoch_loop.py:459]: same ques here

File: pytorch_lightning/loops/epoch/training_epoch_loop.py (hunk 7 above)
Comment (carmocca) [reply pytorch_lightning/loops/epoch/training_epoch_loop.py:456]: This isn't exactly equivalent. We want to sort the keys but keep the values

Whereas your suggestions sorts and keeps the values

File: pytorch_lightning/loops/epoch/training_epoch_loop.py (hunk 9 above)
Comment (carmocca) [reply pytorch_lightning/loops/epoch/training_epoch_loop.py:459]: This calls `len(array)` so I think it's expected to be a List

File: pytorch_lightning/loops/epoch/training_epoch_loop.py (hunk 7 above)
Comment (rohitgr7) [reply pytorch_lightning/loops/epoch/training_epoch_loop.py:456]: oh! okay!

File: pytorch_lightning/loops/epoch/training_epoch_loop.py (hunk 8 above)
Comment (carmocca) [reply pytorch_lightning/loops/epoch/training_epoch_loop.py]: Technically yes, done with `overload`

File: pytorch_lightning/loops/epoch/training_epoch_loop.py [hunk 10]
```diff
@@ -381,6 +394,81 @@ def _save_loggers_on_train_batch_end(self) -> None:
...
+
+
+@overload
+def _recursive_unpad(nested: Any, value: Optional[Any] = None) -> Any:
+    ...
+
+
+@overload
+def _recursive_unpad(nested: List[Any], value: Optional[Any] = None) -> List[Any]:
+    ...
```
Comment (rohitgr7) [inline pytorch_lightning/loops/epoch/training_epoch_loop.py:415]: sorry for nerding out 😅 , but what does it do? never seen this before.

**Review by rohitgr7 — APPROVED** [review]
LG!

File: pytorch_lightning/loops/epoch/training_epoch_loop.py (hunk 10 above)
Comment (carmocca) [reply pytorch_lightning/loops/epoch/training_epoch_loop.py:415]: https://adamj.eu/tech/2021/05/29/python-type-hints-how-to-use-overload/

Comment (rohitgr7) [reply pytorch_lightning/loops/epoch/training_epoch_loop.py:415]: this looks like a good blog to follow .. thanks 👍
